### PR TITLE
[Example] [autodiff] Add ad_gravity.py (compute gradient from potential energy)

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
I :heart: AutoDiff! But currently our AD examples are only about ML... Let's add a CG-related example now.

### CODE REVIEW

File: python/taichi/lang/__init__.py [hunk 1]
```diff
@@ -265,7 +265,13 @@ def svd(A, dt=None):
 
 def Tape(loss, clear_gradients=True):
     get_runtime().materialize()
-    assert loss.snode().ptr.has_grad(), "gradient for loss not allocated"
+    if len(loss.shape) != 0:
+        raise RuntimeError(
+            'The loss tensor for `Tape` must be a 0D tensor, i.e. scalar')
```
Comment (Rullec) [inline python/taichi/lang/__init__.py:270]:
```suggestion
            'The loss of `Tape` must be a 0D tensor, i.e. scalar')
```

File: python/taichi/lang/__init__.py [hunk 2]
```diff
@@ -265,7 +265,13 @@ def svd(A, dt=None):
...
+    if not loss.snode().ptr.has_grad():
+        raise RuntimeError(
+            'Gradient for loss not allocated, please use ti.var(..., needs_grad=True)'
```
Comment (Rullec) [inline python/taichi/lang/__init__.py:273]:
```suggestion
            'Gradients of loss are not allocated, please use ti.var(..., needs_grad=True)'
```

File: examples/ad_gravity.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import taichi as ti
+
```
Comment (Rullec) [inline examples/ad_gravity.py:2]: Just wonder: Dont we need to keep `ti.init` here anymore?

**Comment by Rullec** [comment]
LGTM and few nits!
I don't know whether it's proper to add a reference in `differential_programming.rst`, guiding our reader to find this example.

File: examples/ad_gravity.py (hunk 3 above)
Comment (archibate) [reply examples/ad_gravity.py:2]:
```suggestion
ti.init()

```
Yes, sometimes Taichi developments can break these examples without `ti.init()`.

**Comment by archibate** [comment]
Great! Let's merge this to catch up with the next release. I'll try to complete the doc in AD and add a link to this example later.

**Pull request merged**